lib/lifecycle/property.js

```diff
@@ -0,0 +1,169 @@
+          info.attributeMap = {};
+
+          elem.removeAttribute = function (attrName) {
+            info.updatingAttribute = true;
+            info.removeAttribute.call(this, attrName);
+            if (attrName in info.attributeMap) {
+              elem[info.attributeMap[attrName]] = undefined;
+            }
+            info.updatingAttribute = false;
+          };
+
+          elem.setAttribute = function (attrName, attrValue) {
+            info.updatingAttribute = true;
+            info.setAttribute.call(this, attrName, attrValue);
+            if (attrName in info.attributeMap) {
+              // Could also call getAttribute() but this does the same thing.
+              attrValue = String(attrValue);
+              elem[info.attributeMap[attrName]] = opts.deserialize ? opts.deserialize(attrValue) : attrValue;
+            }
+            info.updatingAttribute = false;
+          };
+        }
+
+        info.attributeMap[info.linkedAttribute] = name;
+      }
+
+      if (initialValue === undefined) {
+        if (info.linkedAttribute && elem.hasAttribute(info.linkedAttribute)) {
+          var attributeValue = elem.getAttribute(info.linkedAttribute);
+          initialValue = opts.deserialize ? opts.deserialize(attributeValue) : attributeValue;
+        } else if (typeof opts['default'] === 'function') {
+          initialValue = opts['default']();
+        } else if (opts['default'] !== undefined) {
+          initialValue = opts['default'];
+        }
+      }
+
+      info.internalValue = initialValue;
+    };
+
+    prop.get = function () {
+      var value = opts.get ? opts.get(this) : (0, _data['default'])(this, 'api/property/' + name).internalValue;
+
+      if (value === undefined && opts['default'] !== undefined) {
+        if (typeof opts['default'] === 'function') {
+          value = opts['default']();
+        } else {
+          value = opts['default'];
+        }
+      }
+
+      return value;
+    };
+
+    prop.set = function (newValue) {
+      var info = (0, _data['default'])(this, 'api/property/' + name);
+
+      if (info.updatingProperty) {
+        return;
+      }
+
+      info.updatingProperty = true;
+      var oldValue = this[name];
+
+      if (opts.type) {
+        newValue = opts.type(newValue);
+      }
+
+      if (!opts.get) {
+        info.internalValue = newValue;
+      }
+
+      if (info.linkedAttribute && !info.updatingAttribute) {
+        var serializedValue = opts.serialize ? opts.serialize(newValue) : newValue;
+        if (serializedValue === undefined) {
+          info.removeAttribute.call(this, info.linkedAttribute);
+        } else {
+          info.setAttribute.call(this, info.linkedAttribute, serializedValue);
+        }
+      }
+
+      var changeData = {
+        name: name,
+        newValue: newValue,
+        oldValue: oldValue
+      };
+
+      if (opts.set) {
+        opts.set(this, changeData);
+      }
+
+      if (opts.emit) {
+        var eventName = opts.emit;
```

Comment:
If `opts.emit !== true` then should we `String(opts.emit)` to get the event name?